### ISSUE TITLE
OMPS NPP NMSO2

### ISSUE THREAD

**Issue by bbakernoaa** [pytroll/satpy#57]
I'm a newer user to satpy and am having trouble loading some OMPS data downloaded from the NASA EarthData website.  The files names are as follows: 

> OMPS-NPP_NMSO2-L2_2017m0707t235247_o29505_2017m0708t125922.h5

I thought that I would need to modify the omps_edr.yaml file to address this but I continue to get an error when trying to load it like so
```python
gs = Scene(platform_name='NPP',sensor='omps',filenames=files)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-2d834f8862df> in <module>()
----> 1 gs = Scene(platform_name='NPP',sensor='OMPS',filenames=files)

/data/aqf2/barryb/anaconda2/lib/python2.7/site-packages/satpy/scene.py in __init__(self, filenames, ppp_config_dir, reader, base_dir, sensor, start_time, end_time, area, reader_kwargs, **metadata)
     99                                                     base_dir=base_dir,
    100                                                     reader=reader,
--> 101                                                     reader_kwargs=reader_kwargs)
    102         self.info.update(self._compute_metadata_from_readers())
    103         self.datasets = DatasetDict()

/data/aqf2/barryb/anaconda2/lib/python2.7/site-packages/satpy/scene.py in create_reader_instances(self, filenames, base_dir, reader, reader_kwargs)
    145                       sensor=self.info.get("sensor"),
    146                       filenames=filenames,
--> 147                       reader_kwargs=reader_kwargs)
    148 
    149     @property

/data/aqf2/barryb/anaconda2/lib/python2.7/site-packages/satpy/readers/__init__.pyc in __call__(self, filenames, sensor, reader, reader_kwargs)
    350                     remaining_filenames)))
    351         if not reader_instances:
--> 352             raise ValueError("No supported files found")
    353         return reader_instances
    354 

ValueError: No supported files found
```

**Comment by davidh-ssec** [comment]
The OMPS reader was created as sort of a proof-of-concept so it may have limited functionality from what you expect. It was also my first time working with OMPS data so I may have done something wrong. As you can see in the omps_edr.yaml file we currently only support the following filename scheme for SO2:

```
OMPS-NPP-TC_EDR_SO2NRT-%Ym%m%dt%H%M%S-o00001-%Ym%m%dt%H%M%S.he5
```

Do you have any idea what the differences are between your file and the above type of file?

**Comment by bbakernoaa** [comment]
I may be wrong but isn't there two types of files in there

```
file_types:
  omps_tc_so2_edr:
    file_reader: !!python/name:satpy.readers.omps_edr.EDREOSFileHandler
    file_patterns: ['{instrument_shortname}-{platform_shortname}-TC_EDR_SO2NRT-{start_time:%Ym%m%dt%H%M%S}-o{orbit:05d}-{end_time:%Ym%m%dt%H%M%S}.he5']
  omps_tc_to3_edr:
    file_reader: !!python/name:satpy.readers.omps_edr.EDRFileHandler
    file_patterns: ['{instrument_shortname}-{platform_shortname}-TC_EDR_TO3-{version}-{start_time:%Ym%m%dt%H%M%S}-o{orbit:05d}-{end_time:%Ym%m%dt%H%M%S}.h5']
```
I had tried to modify it to the following:

```file_types:
  omps_tc_so2_edr:
    file_reader: !!python/name:satpy.readers.omps_edr.EDREOSFileHandler
    file_patterns: ['{instrument_shortname}-{platform_shortname}-TC_EDR_SO2NRT-{start_time:%Ym%m%dt%H%M%S}-o{orbit:05d}-{end_time:%Ym%m%dt%H%M%S}.he5']
  omps_tc_to3_edr:
    file_reader: !!python/name:satpy.readers.omps_edr.EDRFileHandler
    file_patterns: ['OMPS_NPP_NMSO2-L2_{start_time:%Ym%m%dt%H%M%S}_o{orbit:05d}_{end_time:%Ym%m%dt%H%M%S}.h5']
```

Simply opening it with h5py 

```python

In [26]: f['GeolocationData'].keys()
Out[26]: 
[u'GroundPixelQualityFlags',
 u'ImageMidpoint_TAI93',
 u'InstrumentQualityFlags',
 u'Latitude',
 u'LatitudeCorner',
 u'Longitude',
 u'LongitudeCorner',
 u'RelativeAzimuthAngle',
 u'SolarAzimuthAngle',
 u'SolarZenithAngle',
 u'SpacecraftAltitude',
 u'SpacecraftLatitude',
 u'SpacecraftLongitude',
 u'SpacecraftSolarZenith',
 u'UTC_CCSDA_A',
 u'ViewingAzimuthAngle',
 u'ViewingZenithAngle']

In [15]: f['ScienceData'].keys()
Out[15]: 
[u'AerosolIndex',
 u'CloudFraction',
 u'CloudPressure',
 u'CloudRefletivity',
 u'ColumnAmountO3isf',
 u'ColumnAmountO3pair',
 u'ColumnAmountSO2_ABV',
 u'ColumnAmountSO2_PBL',
 u'ColumnAmountSO2_STL',
 u'ColumnAmountSO2_TRL',
 u'ColumnAmountSO2_TRM',
 u'ColumnAmountSO2_TRU',
 u'GroundRefletivity',
 u'PixelQualityFlags',
 u'RadiativeCloudFraction',
 u'ScenePressure',
 u'SceneRefletivity',
 u'SlantColumnAmountSO2',
 u'TerrainPressure']
```

**Comment by davidh-ssec** [comment]
Yes there are two file patterns but only one for SO2. Looking at the yaml file again it looks like the SO2 file is an HDFEOS (HDF5 though?) file. Why? I have no idea.

If the files you use are more common it may be best to use those instead. You could either add a new file type and then a new dataset entry for whatever fields you would like to read or you could modify the exiting SO2 related entries.

**Comment by bbakernoaa** [comment]
@davidh-ssec thank you! I'm trying to figure this out.  I'll take a crack at it

**Comment by bbakernoaa** [comment]
@davidh-ssec 

I've modified the omps_edr.yaml file to allow opening the file but it isn't allowing me to load the data 

```python
gs = Scene(platform_name='NPP',sensor='omps',base_dir='/data/aqf/barryb/OMPS-HI',start_time=datetime(2017,7,1,0,0),end_time=datetime(2017,7,30,0,0))
gs.available_dataset_names()
Out[30]: 
['cldfra',
 'cldpres',
 'cldrefl',
 'latitude_so2',
 'longitude_so2',
 'so2_trm',
 'tcso2_abv',
 'tcso2_pbl',
 'tcso2_stl',
 'tcso2_trl',
 'tcso2_trm',
 'tcso2_tru']
gs.load(['tcso2_pbl'])
print gs['tcso2_pbl']
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-34-42f1efb4a130> in <module>()
----> 1 print gs['tcso2_pbl']

/data/aqf2/barryb/anaconda2/lib/python2.7/site-packages/satpy/scene.pyc in __getitem__(self, key)
    297     def __getitem__(self, key):
    298         """Get a dataset."""
--> 299         return self.datasets[key]
    300 
    301     def __setitem__(self, key, value):

/data/aqf2/barryb/anaconda2/lib/python2.7/site-packages/satpy/readers/__init__.pyc in __getitem__(self, item)
    178         key = self.get_key(item)
    179         if key is None:
--> 180             raise KeyError("No dataset matching '{}' found".format(str(item)))
    181         return super(DatasetDict, self).__getitem__(key)
    182 

KeyError: "No dataset matching 'tcso2_pbl' found"
```

I'm attaching the new omps_edr.yaml along with a single track of the OMPS data 

[omps_testfile_and_yaml.zip](https://github.com/pytroll/satpy/files/1218041/omps_testfile_and_yaml.zip)

Thank you for helping me through this.  I'm excited about getting satpy to work for these applications

**Comment by davidh-ssec** [comment]
@bbakernoaa You are still using the EOS file handler, but your files are not HDF-EOS files. I tried testing your YAML file and found other differences from the HDF5 format that I originally created the reader for. Are you aware of why your files are different? Was there a change in the processing software for these systems? At the moment I don't remember where I got my files from, but if needed we can switch to only supporting your files.

**Comment by bbakernoaa** [comment]
@davidh-ssec I'm not aware why the files are different.  I retrieved the files from https://earthdata.nasa.gov/ 

I quickly tried to change the reader ```to satpy.readers.omps_edr.EDRFileHandler```

is there a generic hdf reader we should try?

**Comment by davidh-ssec** [comment]
@bbakernoaa I'm looking in to the types of files. The EDRFileHandler is based on a semi-generic HDF5 reader, but it expects attributes to be named like the files I was using. The files you are using look more sane to me so I'm fine switching, but first I want to see if I can track down the files I was using and see what's going on.

Differences in the attributes are things like "ValidRange" to "valid_range" and "Units" to "units". So it is possible that there was a change to the processing software to be more CF compliant-ish. I'm not super familiar with OMPS so I'll see what I can find.

**Comment by bbakernoaa** [comment]
@davidh-ssec 

Thank you for looking into this.  I was asked by a colleague to take a look into using OMPS so here I am

**Comment by davidh-ssec** [comment]
Now I'm even more confused, I found this: ftp://omisips1.omisips.eosdis.nasa.gov/OMPS/LANCE/NMSO2-L2-NRT-NRT/OMPS-NPP_NMSO2-L2-NRT_2017m0804t030731_o29890_2017m0804t021637.he5

So same filename scheme as yours, except its HDF-EOS (.he5).

**Comment by davidh-ssec** [comment]
@bbakernoaa I'm having difficulty finding your exact files. Where did you get them from? I know you said EarthData, but when I get to here I don't have any Product links to click except to go to the SIPS: https://earthdata.nasa.gov/earth-observation-data/near-real-time/omps-nrt

**Comment by bbakernoaa** [comment]
https://search.earthdata.nasa.gov/search/granules?projectId=5490949863

**Comment by bbakernoaa** [comment]
To be honest if I know a location where I can get the same data, I don't really care if it is in HDF-EOS or whatever format.  I just want to be able to read and use it.  If satpy can already handle the OMPS HDF-EOS files I don't have an issue using that

**Comment by davidh-ssec** [comment]
@bbakernoaa If you install satpy from the `develop` branch on github your files should now work. I took your code, fixed/modified some of it, and it looks like it works. I created a new file type so that my original files still work. A coworker pointed me to the original site that I downloaded these from which may not be supported anymore (I'm not sure): ftp://omisips1.omisips.eosdis.nasa.gov/OMPS/LANCE/NMSO2-L2-NRT-NRT

It's a little annoying because now we have to have your `tcX` names even though they could probably be just `X`, but can't be because they would conflict with the old dataset definitions. Let me know if you have any issues.
